File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    static int par[], height[];
    
    public int solution(int n, int[][] computers) {
        par = new int[n];
        height = new int[n];
        
        Arrays.fill(par, -1);
        Arrays.fill(height, 1);
        
        for (int i=0;i<n;i++) for (int j=i+1;j<n;j++) {
            if (computers[i][j]==1) {
                merge(i, j);
            }
        }
        
        int answer = 0;
        for (int p:par) {
            if (p==-1) answer++;
        }

        return answer;
    }
    
    private static int find(int k) {
        if (par[k]==-1) return k;
        return par[k] = find(par[k]);
    }
    
    private static void merge(int a, int b){
        a = find(a);
        b = find(b);
        if (a==b) return;
        
        // 트리의 높이가 낮은 쪽을 높은 쪽 아래에 붙임
        if (height[a]>height[b]) {
            par[b] = a;
        }
        else {
            par[a] = b;
            if (height[a]==height[b]) height[b]++;
        }
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <vector>

using namespace std;

// 컴퓨터 번호: 0 ~ n-1
vector<int> par;

// c++은 Java와 달리 함수나 변수를 사용하기 전에 컴파일러가 이름을 먼저 알고 있어야 함(컴파일러가 코드를 위에서 아래로 읽기 때문에)
// 아래쪽에 정의된 함수를 위쪽 코드에서 바로 호출하면 아직 선언되지 않은 함수로 인식되어 컴파일 오류가 남(위쪽에 바로 구현까지 해도 ㄱㅊ)
int find(int a);

void merge(int a, int b);

int solution(int n, vector<vector<int>> computers) {
    par.resize(n);
    
    for (int i=0; i<n; i++) {
        par[i] = -1;
    }
    
    for (int i=0; i<computers.size(); i++) {
        for (int j=i+1; j<computers[i].size(); j++) {
            if (computers[i][j]==1) {
                merge(i, j);
            }
        }
    }
    
    int answer = 0;
    for (int i=0; i<n; i++) {
        if (par[i] == -1) {
            answer++;
        }
    }
    
    return answer;
}

int find(int a) {
    if (par[a]==-1) {
        return a;
    }
    
    return par[a] = find(par[a]);
}

void merge(int a, int b) {
    a = find(a);
    b = find(b);
    
    if (a == b) {
        return;
    }
    
    par[a] = b;
}